### PR TITLE
chore: update DataFusion and arrow / arrow-flight / parquet to `39.0.0`

### PR DESCRIPTION
**Pull request by appletreeisyellow**
# Rationale:

* Keep up with dependencies
* Fix breaking changes:
    * Reflect the refactor work https://github.com/apache/arrow-rs/pull/4116
    * Update CSV schema https://github.com/apache/arrow-rs/pull/4133

# Changes:

1. Update DataFusion pin to https://github.com/apache/arrow-datafusion/commit/e6d7e46dedbe5046e4606bfd3d7a1199dd0aaae2
2. Update arrow / arrow-flight / parquet to `39.0.0`

### CODE REVIEW

File: query_functions/src/coalesce_struct.rs [hunk 1]
```diff
@@ -148,8 +148,7 @@ fn arrow_coalesce_struct(
             .zip(array1.fields())
             .map(|((col1, col2), field)| {
                 let out = arrow_coalesce_struct(&col1, &col2)?;
-                // TODO: avoid field clone once https://github.com/apache/arrow-rs/pull/4116 is available
-                Ok((field.as_ref().clone(), out)) as Result<_, DataFusionError>
+                Ok((Arc::<Field>::clone(field), out)) as Result<_, DataFusionError>
```
Comment (appletreeisyellow) [inline query_functions/src/coalesce_struct.rs:151]: @crepererum  I'm not sure if this is the correct way to reflect the changes you made in https://github.com/apache/arrow-rs/pull/4116. I'm open to change if there is a better way

**Review by tustvold — APPROVED** [review]

File: iox_query/src/exec/seriesset/converter.rs [hunk 2]
```diff
@@ -1233,22 +1233,11 @@ mod tests {
...
-            data.as_bytes(),
-            schema,
-            has_header,
-            delimiter,
-            batch_size,
-            bounds,
-            projection,
-            datetime_format,
-        );
+        let mut reader = csv::ReaderBuilder::new(schema)
```
Comment (alamb) [inline iox_query/src/exec/seriesset/converter.rs:1237]: that is certainly nicer

**Review by alamb — APPROVED** [review]
Lovely -- thank you @appletreeisyellow

**Review by alamb — APPROVED** [review]
Lovely -- thank you @appletreeisyellow

File: query_functions/src/coalesce_struct.rs (hunk 1 above)
Comment (crepererum) [reply query_functions/src/coalesce_struct.rs:151]: That's correct. Although I think you don't need to type-annotate `Arc`, so the following should be enough:

```rust
Ok((Arc::clone(field), out)) as Result<_, DataFusionError>
```

**Review by crepererum — APPROVED** [review]
Small (optional) clean-code nitpick, looks good otherwise. Thanks for taking care of it :)

**Pull request merged**